src/components/cursor-controller.js

```diff
@@ -0,0 +1,393 @@
+          this.currentDistanceMod = 0;
+        } else {
+          this.currentDistance = this.data.maxDistance;
+        }
+      }
+
+      if (this.isGrabbing || !intersection) {
+        const max = Math.max(this.data.minDistance, this.currentDistance - this.currentDistanceMod);
+        const distance = Math.min(max, this.data.maxDistance);
+        this.currentDistanceMod = this.currentDistance - distance;
+        direction.multiplyScalar(distance);
+        point.addVectors(origin, direction);
+        this.data.cursor.object3D.position.copy(point);
+      }
+
+      //update currentTargetType
+      if (this.isGrabbing && !intersection) {
+        this.currentTargetType = TARGET_TYPE_INTERACTABLE;
+      } else if (intersection) {
+        if (this._isClass("interactable", intersection.object.el)) {
+          this.currentTargetType = TARGET_TYPE_INTERACTABLE;
+        } else if (this._isClass("ui", intersection.object.el)) {
+          this.currentTargetType = TARGET_TYPE_UI;
+        }
+      } else {
+        this.currentTargetType = TARGET_TYPE_NONE;
+      }
+
+      //update cursor material
+      const isTarget = this._isTargetOfType(TARGET_TYPE_INTERACTABLE_OR_UI);
+      if ((this.isGrabbing || isTarget) && !wasIntersecting) {
+        wasIntersecting = true;
+        this.data.cursor.setAttribute("material", { color: this.data.cursorColorHovered });
+      } else if (!this.isGrabbing && !isTarget && wasIntersecting) {
+        wasIntersecting = false;
+        this.data.cursor.setAttribute("material", { color: this.data.cursorColorUnhovered });
+      }
+
+      //update line
+      if (this.hasPointingDevice) {
+        this.el.setAttribute("line", { start: origin.clone(), end: this.data.cursor.object3D.position.clone() });
+      }
+    };
+  })(),
+
+  _isClass: function(className, el) {
+    return (
+      el.className === className ||
+      (el.parentNode && el.parentNode != el.sceneEl && this._isClass(className, el.parentNode))
+    );
+  },
+
+  _isTargetOfType: function(mask) {
+    return (this.currentTargetType & mask) === this.currentTargetType;
+  },
+
+  _startTeleport: function() {
+    const hideCursor = !(this.hasPointingDevice || this.inVR);
+    if (this.controller != null) {
+      this.controller.emit(this.data.teleportEvent, {});
+    } else if (this.inVR) {
+      this.data.gazeTeleportControls.emit(this.data.teleportEvent, {});
+    }
+    this.el.setAttribute("line", { visible: hideCursor });
+    this.data.cursor.setAttribute("visible", hideCursor);
+  },
+
+  _endTeleport: function() {
+    const showLine = this.hasPointingDevice || this.inVR;
+    if (this.controller != null) {
+      this.controller.emit(this.data.teleportEndEvent, {});
+    } else if (this.inVR) {
+      this.data.gazeTeleportControls.emit(this.data.teleportEndEvent, {});
+    }
+    this.el.setAttribute("line", { visible: showLine });
+    this.data.cursor.setAttribute("visible", true);
+  },
+
+  _handleMouseDown: function() {
```

Comment:
Might be nicer to just have this method emit the `conterollerEvent` on itself so that everything funnels through `_handleControllerEvent`. This maps closer to where we would eventually want to be where `aframe-input-mappings` is handling mouse events. Depends on how divergent we see mouse clicks being from other input devices.